### ISSUE TITLE
[feature] Integrate a Moon tool for updating content in existing files

### ISSUE THREAD

**Issue by Jad31** [moonrepo/moon#1026]
**Is your feature request related to a problem? Please describe.**

I'd like the capability to update the content of a file within a Moon workspace.

**Describe the solution you'd like**

Introduce a tool similar to the code generation tool that's already available in Moon, but specifically for updating existing file content.

**Describe alternatives you've considered**

**Additional context**

Here's a potential use case:

Suppose we generate an Angular library that either serves as a route or is utilized by an existing Angular application within the workspace. This would require us to update the `tsconfig.json` of the project housing the Angular application. Additionally, the code within the Angular application would need adjustments to properly integrate with the newly generated library.

**Comment by sybernatus** [comment]
I agree, as a mono repo manager, it should be great to be able to update a config file while generating it, like add an angular components, add dependencies (in json, toml, etc...)

I also have this need 🙂

**Comment by milesj** [comment]
Would this be like an interactive editor (vim style), or more like a diffing system where you choose which part to accept, or something else? An example reference would be nice.

**Comment by sybernatus** [comment]
I think @Jad31 is talking more like an Nx style update. 
You can retrieve the content of a config file and update it with a new line. 

For example, the generator add a new JS app (or whatever the technology), then you can:
- retrieve the content of the tsconfig file and update it with your new module
- retrieve the package.json file and add a new script to run it

**Comment by Jad31** [comment]
Nx utilizes a virtual file system tree known as the "Tree". With this system, developers have the capability to execute programmatic manipulations on files within a workspace. The idea behind Nx's Tree is to streamline operations in a monorepo setup by offering tools for dynamic code updates.

When implementing changes or adding new modules with Nx, developers typically:

1. Specify the path to the relevant files, such as `tsconfig.json` or `package.json`.
2. Use methods like `tree.read(filePath)` to retrieve the content of these files.
3. Programmatically make necessary changes, like inserting new scripts, dependencies, or module references.

For Moon, the suggestion is to introduce a similar tool but one that's narrowly tailored: a tool primarily for updating existing file content, without incorporating the broader code-generation capabilities of Nx.

For Moon, the suggestion is to introduce a similar tool but focusing on the update of existing files without the full code generation part that is already present in Moon.

For more details on Nx's Tree and its capabilities, you can refer to their official [documentation](https://nx.dev/packages/devkit/documents/Tree).

**Comment by milesj** [comment]
Ok that makes sense.

The major issue I can see is that moon is Rust based, so providing a programmatic approach isn't simple without using WASM, but that's a lot to throw onto the consumer. I'd have to think about this.

**Comment by svallory** [comment]
@Jad31 and @milesj this can already be achieved for yaml and json files due to the generator ability of merging them.

In fact, [I'm using it here](https://github.com/svallory/moon-launch/blob/main/templates/tools/husky/package.json) to create templates for adding tools to a monorepo. The `moon-launch/tools/husky` template defines a `package.json` with only:

```
{
  "scripts": {
    "prepare": "husky"
  },
  "devDependencies": {
    "husky": "^9.0.11"
  }
}
```

And that gets merged into an existing `package.json` **IF** the user selects `merge`. So there are few things that would need to be done here to improve this:

- Allow a `merge` option in the frontmatter, similar to `force`
- When `merge == true`, moon should give the option to "replace existing file" 
- Ideally, if `merge` is true but `force` is false, moon would show a diff and ask if the user accepts the changes but never give the option to `replace`

**Comment by svallory** [comment]
> For Moon, the suggestion is to introduce a similar tool but one that's narrowly tailored: a tool primarily for updating existing file content, without incorporating the broader code-generation capabilities of Nx.

I'm against this course of action. Nx has a very powerful generation tool, but it's a beast to learn. I have done much more with moon's simple template system in the last 2 weeks than I ever did with nx, because of the necessary time investment.

> The major issue I can see is that moon is Rust based, so providing a programmatic approach isn't simple

I would question if the generator shouldn't be done in typescript instead of Rust. Here are some things to think about:

- The generator does not have the same performance requirements other moon features have
- Its functionality is very separate from the rest
- It's VERY hard to write templates for complex use cases (editing existing file, for example)
- Rust's memory management will make it hard for a lot of features that make template easier

For now, though, how hard would it be to allow a template to simply execute a command and forward its output? (bonus points if I can use template variables when setting the command)

That way one can delegate tasks to a script and use whatever they feel more comfortable with. I think it's the best course to do tasks like updating files, removing files or config, doing stuff based on project structure or features, etc

**Comment by sybernatus** [comment]
> I'm against this course of action. Nx has a very powerful generation tool, but it's a beast to learn. I have done much more with moon's simple template system in the last 2 weeks than I ever did with nx, because of the necessary time investment.

I agree with that, Nx is not well enought documented and the feature is directly included to the generators itself. That's why we are talking about creating a side tool for this purpose and not directly include this feature in generators. I totally agree that today Nx gave me a lot of time to handle it.

In comparison, Moon is so well documented that I prefer use it now as it's save me a lot of time and has much better performance.

---

The need here, is to be able to update different types of config files that is not included in your generators but in the monorepo itself: 

Imagine you are generating a Rust library using moon generator
- You want then add this library to the config.toml workspace to link your library automatically to your rust project
- Also you want to add tasks concerning this library to your task runner config (makefile.toml, package.json, etc.)
Today to have this kind of behavior, you need an extra tool. The idea could be to add a new moon tool directly to just sequencing the 3 tasks.

**Comment by svallory** [comment]
> That's why we are talking about creating a side tool for this purpose and not directly include this feature in generators.

I understand it better now, thanks! Still, even if it is a separate tool, if it's under the umbrella of moon, at least to me, it makes little difference it is another binary or something isolated in a subcommand of moon

> Imagine you are generating a Rust library using moon generator

I understand the use case and I had to do something very similar. This is a template to inject husky config in `package.json`

<img width="648" alt="image" src="https://github.com/moonrepo/moon/assets/117560/baf42876-e22a-4c2c-9b74-22399561eb7b">

I don't know what other languages moon is able to merge though. 

In a talk with @milesj he said it is not the goal of moon generator to be a full-blown generator with all features and I understand that. It's not the focus of moon, and it is challenging to build a good generator. 

So I started working on a fork of hygen called [hypergen](/svallory/hypergen) (veeeery creative name, I know! 😬). I've decided to bring it to parity with moon generator in terms of features, and still keep the extra ones. Also, since it is written in typescript, it will be very easy to support extensions (at least compared to rust)

If you want to join forces there, the help will be much appreciated! :) My goal is to use it to finish [moon-launch](https://github.com/svallory/moonlaunch) which will help every one that uses moon to manage their monorepo config.

But, if you want to stick to rust, the creator of [hygen](/jhondot/hygen) is working on [rrgen](/jondot/rrgen) which is written in rust. And given the quality of his work in hygen, I'm sure it will be awesome.

**Comment by call-me-q** [comment]
Hi, I don't know if this is considered as necroposting or not, but I'm new with `moonrepo` and i fell in love for three reasons :

- Multiple language in monorepo: before the idea of multiple language used in monorepo means I have to dabble with Bazel which well... I have  skill issues.
- Proto : I'm a freelancer-ish guy, I don't know the correct term but basically I join a small company that needs to build a team of developers and help the company gain its track. `proto` amazingly helpful for the IT Support in installing the tools that the devs need. No more maintaining `.sh` to install languages to the company's PC for them.
- Template: To create a project, I don't need to install another package, all I have to do is do `moon generate` and voila~ it's done.

However, It feels like the templating feature needs a bit of enhancement, I can't `inject` or `update content` in existing files.

a simple use case is this:

I create a command that will make a fastify project, this is taken from [this repo with a bit of modification to simplify things](https://github.com/alvfork/fastify-boilerplate/).

`/src/main.ts`
```typescript
import Fastify from "fastify";

const service = Fastify({
    logger: {
        name: "fastify-boilerplate",
        level: "trace",
    },
    genReqId: () => new ShortUniqueId({ length: 10 }).randomUUID(),
});

const start = async () => {
    try {
        service.listen({ port: 1234, host: 'localhost' }, (err) => {
            if (err) {
                console.error(err);
                process.exit(1);
            }
        });
    } catch (err) {
        service.log.error(err);
        process.exit(1);
    }
};

start();
```

And then I would add some routes. It should generate new files like this

`/routes/hello.ts`
```typescript
import type { FastifyInstance } from "fastify";
import eventRoutes from "./events.router";

export default async function (service: FastifyInstance) {
    service.register((server: FastifyInstance) => server.get("/ping", async () => "pong 🏓"), { prefix: "/events" });
}
```

And then inject the `/src/main.ts` with this lines

```typescript
import helloRoute from "./routes/hello";
```

on top of the file and 
```typescript
service.register(helloRoute);
```
after 

```typescript
const service = Fastify({
    logger: {
        name: "fastify-boilerplate",
        level: "trace",
    },
    genReqId: () => new ShortUniqueId({ length: 10 }).randomUUID(),
});
```

and the end result for `/src/main.ts` would be

```typescript
import Fastify from "fastify";
import helloRoute from "./routes/hello";

const service = Fastify({
    logger: {
        name: "fastify-boilerplate",
        level: "trace",
    },
    genReqId: () => new ShortUniqueId({ length: 10 }).randomUUID(),
});

service.register(helloRoute);

const start = async () => {
    try {
        service.listen({ port: 1234, host: 'localhost' }, (err) => {
            if (err) {
                console.error(err);
                process.exit(1);
            }
        });
    } catch (err) {
        service.log.error(err);
        process.exit(1);
    }
};

start();
```

The current problem is, `moon` can't inject the new lines, at least as far as I know. As a work around I use [Scaffdog](https://scaff.dog/) which already provide the feature [here](https://scaff.dog/docs/templates/injection). But it's not having its own weakness, it can't put the templates on `git` repositories. 

Thus, as for now I might use `moonrepo` templating for project scaffolding while using `scaffdog` for creating the detailed generation.

The reason why I'm writing this is to **plead** the contributors, maintainers, authors to enhance the templating feature although in previous post I saw this post from @svallory 

>In a talk with @milesj he said it is not the goal of moon generator to be a full-blown generator with all features and I understand that. It's not the focus of moon, and it is challenging to build a good generator.

I'm not an expert on programming although I've been coding for years (again, skill issue). However, `moonrepo` seems like answering the needs that I have at work. But despite of everything, moonrepo is truly a better way to manage codebases, save developer time, and boost your business.

Thank you for maintaining `moonrepo`.

**Comment by milesj** [comment]
@Jad31 Thanks for all the kind words! moon will continue to get better over time.

As for your request, this is difficult, at least on the Rust side. Ideally, this kind of operation requires parsing the code into an AST, and then modifying the AST to inject the new pieces. Supporting an AST parser for all kinds of file types just isn't worth it in moon/Rust.

However, doing a find and inject based on string/regex matching would be doable, albeit, far less powerful.

**Comment by call-me-q** [comment]
@milesj haha I assume that you're replying to me (hopefully I'm right because if it's not... that's worthy of screenshot in hall of shame).

> As for your request, this is difficult, at least on the Rust side. Ideally, this kind of operation requires parsing the code into an AST, and then modifying the AST to inject the new pieces. Supporting an AST parser for all kinds of file types just isn't worth it in moon/Rust.
> 
> However, doing a find and inject based on string/regex matching would be doable, albeit, far less powerful.

 I think it's fine. While AST is much more powerful, the needs to have this feature outweigh the superiority of AST. Using code generation to scaffold your project seems minor to many experts, but for fresh-grads, hobbyists, and small ad-hoc teams definitely needs this.
 
 I can imagine the future where this feature is finally matured (using AST or not) where the templates are available everywhere. 
 
Use cases:
- a teacher / lecturer giving an assignment "Build an app using node and go, please use MVC architecture and docker". The students able to adapt and learn faster since there's helper and not just "copy paste".
- a request from a client "Can you make me an ERP", the team decided to use DDD and using rust. Instead of spending time on managing the architecture, the team can focus on building the app more.

The best part is, most of the repo management and app building is using moon and proto. No need to argue whether we have to tools A or B since most are included.

**Comment by milesj** [comment]
I don't plan to support this in core. Can maybe be done with an extension plugin.
